### ISSUE TITLE
Header parsing causing invalid memory reference

### ISSUE THREAD

**Issue by dev-embedthis** [embedthis/goahead#285]
### Overview

A security vulnerability affecting GoAhead versions up to 4.1.0 has been identified. This bulletin discusses this flaw and its implications.

### Summary

An invalid set of HTTP headers may trigger an assertion and possible out of bounds memory access. 

### Description

The following request will trigger a memory assertion and subsequent out of bounds memory access. Depending on the system, compiler, operating environment and configuration, this could lead corrupted execution and a denial of service.

```
GET / HTTP/1.0
:
Content-Length:10000
0

```

### Threat Scope

Versions up to and including 4.1.0.

### Remedy

Upgrade to 4.1.1 when released or apply the following patch:

```
diff --git a/src/http.c b/src/http.c
index f89c0d4b..039dc9ca 100644
--- a/src/http.c
+++ b/src/http.c
@@ -1173,7 +1173,10 @@ static void parseHeaders(Webs *wp)
-        assert(bufLen(&wp->rxbuf) >= 2);
+        if (bufLen(&wp->rxbuf) < 2 || wp->rxbuf.servp[0] != '\r' || wp->rxbuf.servp[1] != '\n') {
+            websError(wp, HTTP_CODE_BAD_REQUEST | WEBS_CLOSE, "Bad header termination");
+            return;
+        }
         wp->rxbuf.servp += 2;
     }
```

Please contact Embedthis if you require further information, test code or assistance at dev@embedthis.com.

### References

GoAhead Issue https://github.com/embedthis/goahead/issues/285

### CVE

CVE pending.

#### Thanks

Thanks to Yakun Zhang and Zheng Huang of Baidu Security Lab for finding and diagnosing this issue.

**Comment by theonlydso** [comment]
The given patch applies to 4.1.1. Is there a patch available for 2.1.6 or 2.1.8?

**Comment by dev-embedthis** [comment]
No, version 2.X is obsolete. This issue does apply to version 2.X and there are several other significant security issues in GoAhead 2 that have been addressed in GoAhead 3/4. 

While the GoAhead 3/4 APIs are different in syntax, the underlying functionality is the same, the upgrade to version 4 is quite easy.

Lastly, if you are using version 2, you cannot apply code or patches from the version 4 code base as the licenses are incompatible. You must either adopt a commercial license for version 4 or the GPL. You cannot extract patches from this repository and apply to GoAhead 2 code bases.
